README.md
# web-push

[![Hackage](https://img.shields.io/hackage/v/web-push.svg)](https://hackage.haskell.org/package/web-push)
[![Spec and Mock Tests](https://github.com/cotrone/web-push/actions/workflows/web-push-tests.yml/badge.svg)](https://github.com/cotrone/web-push/actions/workflows/web-push-tests.yml)
[![Example Browser Tests](https://github.com/cotrone/web-push/actions/workflows/web-push-example.yml/badge.svg)](https://github.com/cotrone/web-push/actions/workflows/web-push-example.yml)

Send web push notifications to browsers

Forked from https://github.com/sarthakbagaria/web-push

## Usage

Guides to using Web Push API in browsers can be found on [Mozilla's](https://developer.mozilla.org/en/docs/Web/API/Push_API) and [Google's](https://developers.google.com/web/fundamentals/engage-and-retain/push-notifications/) docs, or you can check out the demo app in the example folder. To run the demo app:

with nix: `nix run .#web-push-example:exe:web-push-example-server`

with cabal: `cabal run web-push-example-server`

stack: `stack exec web-push-example-server`

Then access `localhost:3000` from a browser

For production use, store a set of VAPID keys securely and use them for all push notification subscriptions and messages; public key will have to be exposed to client's browser when subscribing to push notifications, but private key must be kept secret and used when generating push notifications on the server. If VAPID keys are re-generated, all push notifications will require re-subscriptions. Also save the latest subscription details such as endpoint from user's browser session securely in the database and use them to send push notifications to the user later.

## Build

The library builds and tests run with both cabal and stack.

There is also a nix flake setup to manage the dependencies for tests.

To build a cabal environment for the project that

### Nix

The nix flake includes:

- A development environment for web-push and web-push-example through `nix develop`
- A package for the example server `nix run .#web-push-example-server`
- A package for the mock web push server from https://github.com/marc1706/web-push-testing
  - `nix run .#web-push-testing -- start` and `nix run .#web-push-testing -- stop` to control it
  - Launch the server directly with `nix run .#web-push-testing-server -- 8090`
- Tests for `web-push` and `web-push-example`
  - For `web-push`: `nix run .#web-push-test`
  - For `web-push-example`: `nix run .#web-push-example-test`
- Browser tests for the `web-push-example` server with `nix run .#web-push-example-test`

## Example

There is a minimal example web application that generates keys, stores subscriptions, and sends notifications to all subscribers in [web-push-example](web-push-example/README.md)

## References

Current implementation is based on the following versions of the drafts:
- [https://tools.ietf.org/html/draft-ietf-webpush-encryption-04](https://tools.ietf.org/html/draft-ietf-webpush-encryption-04)
- [https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-02](https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-02)
- [https://tools.ietf.org/html/draft-ietf-webpush-protocol-10](https://tools.ietf.org/html/draft-ietf-webpush-protocol-10)
- [https://tools.ietf.org/html/draft-ietf-webpush-vapid-01](https://tools.ietf.org/html/draft-ietf-webpush-vapid-01)


web-push-example/CHANGELOG.md
# Revision history for web-push-example

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Main.hs
module Main where

import WebPushEncryptionSpec
import WebPushMock
import Test.Tasty
import Test.Tasty.Hspec

main :: IO ()
main = do
  encryptionSpec <- testSpecs spec
  defaultMain $
    testGroup "Tests" [
        testGroup "WebPushEncryptionSpec" encryptionSpec
      , testGroup "Mock" [testSendMessage]
    ]

test/WebPushMock.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
module WebPushMock where

import           Control.Concurrent
import           Control.Exception
import           Control.Lens               hiding ((.=))
import           Data.Aeson
import           Data.ByteString            (ByteString)
import qualified Data.ByteString            as BS
import           Data.ByteString.Base64.URL
import qualified Data.ByteString.Lazy       as BSL
import           Data.Text                  (Text)
import qualified Data.Text                  as Text
import qualified Data.Text.Encoding         as TE
import           Data.Time
import qualified Network.HTTP.Client        as HTTP
import           Network.Wreq
import           System.Process
import           Test.Tasty
import           Test.Tasty.HUnit
import           Web.WebPush

{-
  These are tests to be ran against a mock web-push server from https://github.com/marc1706/web-push-testing
-}

testSendMessage :: TestTree
testSendMessage =
  localOption (mkTimeout (10 ^ (7 :: Integer))) $ withResource initWebPushTestingServer terminateProcess $ \_ -> do
    testCaseSteps "Mock subscription" $ \step -> do
      step "Waiting for status to be accessible"
      waitForStatus

      keys <- either fail (pure . readVAPIDKeys) =<< generateVAPIDKeys
      publicKeyBytes <- either fail (pure . BS.pack) $ vapidPublicKeyBytes keys
      let subscriptionOptions = SubscribeOptions True publicKeyBytes

      step "Creating a test subscription to the mock server"
      subscribeResponse <- webPushSubscribe subscriptionOptions
      let subscription = subscribeResponse ^. responseBody

      step "Sending message through the mock server"
      time <- getCurrentTime
      let notification = (mkPushNotification (endpoint subscription) (p256dh subscription) (auth' subscription))
            & pushExpireInSeconds .~ 60 * 60 * 12
            & pushMessage .~ message
            & pushSenderEmail .~ "test@example.com"
          message :: Value
          message = object [
              "title" .= ("Web Push Test" :: Text)
            , "body" .= ("Hello" :: Text)
            , "icon" .= ("" :: Text)
            , "tag" .= Text.pack (show time)
            , "url" .= ("http://localhost:3000" :: Text)
            ]
          encodedMessage = TE.decodeUtf8 $ BSL.toStrict $ encode message
      endpointManager <- HTTP.newManager HTTP.defaultManagerSettings
      either (assertFailure . show) (const (pure ())) =<< sendPushNotification keys endpointManager notification

      step "Getting notifications"
      notificationsResponse <- webPushGetNotificationsFor (clientHash subscription)
      let notifications = notificationsResponse ^. responseBody . to messages
      assertEqual "One notification" 1 (length notifications)
      assertEqual "Notification content" encodedMessage (head notifications)

webPushTestingPort :: Int
webPushTestingPort = 8090

webPushTestingUrl :: String
webPushTestingUrl = "http://localhost:" <> show webPushTestingPort

-- Initializes the web push testing server binary from path
initWebPushTestingServer :: IO ProcessHandle
initWebPushTestingServer = do
  (_stdIn, _stdOut, _stdErr, procHandle) <- createProcess (proc "web-push-testing-server" [show webPushTestingPort])
  pure procHandle

waitForStatus :: IO ()
waitForStatus = do
  resp :: Either SomeException (Response BSL.ByteString) <- try webPushStatus
  case resp of
    Left err -> do
      putStrLn $ "Waiting for web-push-testing-server to start: " <> show err
      threadDelay 1000000
      waitForStatus
    Right _ -> pure ()

{-
API for the mock server
-}
webPushStatus :: IO (Response BSL.ByteString)
webPushStatus = post (webPushTestingUrl <> "/status") (mempty :: ByteString)

newtype ClientHash = ClientHash {
  unClientHash :: Text
} deriving (Eq, Ord, Show)

instance ToJSON ClientHash where
  toJSON (ClientHash hash) = object ["clientHash" .= hash]

data SubscribeOptions = SubscribeOptions {
  userVisibleOnly :: Bool,
  applicationServerKey :: ByteString
} deriving (Eq, Ord, Show)

instance ToJSON SubscribeOptions where
  toJSON opts = object [
      "userVisibleOnly" .= stringlyBool (userVisibleOnly opts)
    , "applicationServerKey" .= encodeBase64Unpadded (applicationServerKey opts)
    ]
    where
      stringlyBool :: Bool -> Text
      stringlyBool True = "true"
      stringlyBool False = "false"

data SubscriptionResult = SubscriptionResult {
  endpoint :: Text,
  p256dh :: Text,
  auth' :: Text,
  clientHash :: ClientHash
} deriving (Eq, Ord, Show)

instance FromJSON SubscriptionResult where
  parseJSON = withObject "SubscriptionResult" $ \o -> do
    dataObj <- o .: "data"
    SubscriptionResult
      <$> dataObj .: "endpoint"
      <*> (dataObj .: "keys" >>= (.: "p256dh"))
      <*> (dataObj .: "keys" >>= (.: "auth"))
      <*> (ClientHash <$> dataObj .: "clientHash")

webPushSubscribe :: SubscribeOptions -> IO (Response SubscriptionResult)
webPushSubscribe opts = asJSON =<< post (webPushTestingUrl <> "/subscribe") (toJSON opts)

data GetNotifications = GetNotifications {
  messages :: [Text]
} deriving (Eq, Ord, Show)

instance FromJSON GetNotifications where
  parseJSON = withObject "GetNotifications" $ \o -> do
    d <- o .: "data"
    GetNotifications <$> d .: "messages"

webPushGetNotificationsFor :: ClientHash -> IO (Response GetNotifications)
webPushGetNotificationsFor client = asJSON =<< post (webPushTestingUrl <> "/get-notifications") (toJSON client)

test/WebPushEncryptionSpec.hs
{-# LANGUAGE OverloadedStrings #-}

module WebPushEncryptionSpec where

import Web.WebPush.Internal

import Test.Hspec

import qualified Data.Binary                     as Binary
import Data.ByteString.Base64.URL
import Data.ByteString                                         (ByteString)
import qualified Data.ByteString                 as BS
import qualified Data.ByteString.Lazy            as LB


-- Test to match input and output from the specification.
-- https://tools.ietf.org/html/draft-ietf-webpush-encryption-04
spec :: Spec
spec = describe "Web Push Encryption Test" $ do

    let encryptionInput = EncryptionInput
          { applicationServerPrivateKey = bytes32Int $ bsTo32Bytes $ BS.concat [ "nCScek-QpEjmOOlT-rQ38nZ"
                                                                               , "zvdPlqa00Zy0i6m2OJvY"
                                                                               ]
          , userAgentPublicKeyBytes = decodeBase64Lenient $ BS.concat [ "BCEkBjzL8Z3C-oi2Q7oE5t2Np-"
                                                                        , "p7osjGLg93qUP0wvqR"
                                                                        , "T21EEWyf0cQDQcakQMqz4hQKYOQ3il2nNZct4HgAUQU"
                                                                        ]
          , authenticationSecret = decodeBase64Lenient "R29vIGdvbyBnJyBqb29iIQ"
          , salt = decodeBase64Lenient "lngarbyKfMoi9Z75xYXmkg"
          , plainText = "I am the walrus"
          , paddingLength = 0
          }

        encryptionOutput = webPushEncrypt encryptionInput

        expectedEncryptionOutput = EncryptionOutput
            { sharedECDHSecretBytes = decodeBase64Lenient $ BS.concat [ "RNjC-"
                                                                        -- NOTE: the specs example might have printed this wrong
                                                                        -- there should be two consecutive hyphens and not one
                                                                        -- near the end of the encoded string, just before "NOQ6Y"
                                                                        , "NVW4BGJbxWPW7G2mowsLeDa53LYKYm4--NOQ6Y"
                                                                        ]
            , inputKeyingMaterialBytes = decodeBase64Lenient $ BS.concat [ "EhpZec37Ptm4IRD5-jtZ0q6r1iK5vYmY1tZwtN8"
                                                                           , "fbZY"
                                                                           ]
            , contentEncryptionKeyContext = decodeBase64Lenient $ BS.concat [ "Q29udGVudC1FbmNvZGluZzogYWVzZ2NtAFAtMjU2AABB BCEkBjzL8Z3C-"
                                                                              , "oi2Q7oE5t2Np-p7osjGLg93qUP0wvqR"
                                                                              , "T21EEWyf0cQDQcakQMqz4hQKYOQ3il2nNZct4HgAUQUA"
                                                                              , "QQTaEQ22_OCRpvIOWeQhcbq0qrF1iddSLX1xFmFSxPOW"
                                                                              , "OwmJA417CBHOGqsWGkNRvAapFwiegz6Q61rXVo_5roB1"
                                                                              ]
            , contentEncryptionKey = decodeBase64Lenient "AN2-xhvFWeYh5z0fcDu0Ww"
            , nonceContext = decodeBase64Lenient $ BS.concat [ "Q29udGVudC1FbmNvZGluZzogbm9uY2UAUC0yNT"
                                                               , "YAAEEE ISQGPMvxncL6iLZDugTm3Y2n6nuiyMYuD3epQ_TC-pFP"
                                                               , "bUQRbJ_RxANBxqRAyrPiFApg5DeKXac1ly3geABRBQBB"
                                                               , "BNoRDbb84JGm8g5Z5CFxurSqsXWJ11ItfXEWYVLE85Y7"
                                                               , "CYkDjXsIEc4aqxYaQ1G8BqkXCJ6DPpDrWtdWj_mugHU"
                                                               ]
            , nonce = decodeBase64Lenient "JY1Okw5rw1Drkg9J"
            , paddedPlainText = decodeBase64Lenient "AABJIGFtIHRoZSB3YWxydXM"
            , encryptedMessage = decodeBase64Lenient $ BS.concat [ "6nqAQUME8hNqw5J3kl8cpVVJylXKYqZOeseZG8UueKpA" ]
            }

    it "should match the shared secret" $ do
        (sharedECDHSecretBytes <$> encryptionOutput) `shouldBe` Right (sharedECDHSecretBytes expectedEncryptionOutput)

    it "should match the input keying material" $ do
        (inputKeyingMaterialBytes <$> encryptionOutput) `shouldBe` Right (inputKeyingMaterialBytes expectedEncryptionOutput)

    it "should match the context for content encryption key derivation" $ do
        (contentEncryptionKeyContext <$> encryptionOutput) `shouldBe` Right (contentEncryptionKeyContext expectedEncryptionOutput)

    it "should match the content encryption key" $ do
        (contentEncryptionKey <$> encryptionOutput) `shouldBe` Right (contentEncryptionKey expectedEncryptionOutput)

    it "should match the context for nonce derivation" $ do
        (nonceContext <$> encryptionOutput) `shouldBe` Right (nonceContext expectedEncryptionOutput)

    it "should match the nonce" $ do
        (nonce <$> encryptionOutput) `shouldBe` Right (nonce expectedEncryptionOutput)

    it "should match the paddded plain text" $ do
        (paddedPlainText <$> encryptionOutput) `shouldBe` Right (paddedPlainText expectedEncryptionOutput)

    it "should match the encrypted message" $ do
        (encryptedMessage <$> encryptionOutput) `shouldBe` Right (encryptedMessage expectedEncryptionOutput)



    where
        bsTo32Bytes :: ByteString -> Bytes32
        bsTo32Bytes = Binary.decode . LB.fromStrict . decodeBase64Lenient


src/Web/WebPush/Internal.hs
{-# LANGUAGE RecordWildCards, OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Web.WebPush.Internal where

import           Control.Monad.IO.Class     (MonadIO, liftIO)
import           Crypto.Cipher.AES          (AES128)
import qualified Crypto.Cipher.Types        as Cipher
import qualified Crypto.ECC
import           Crypto.Error               (CryptoError, eitherCryptoError)
import           Crypto.Hash.Algorithms     (SHA256 (..))
import qualified Crypto.MAC.HMAC            as HMAC
import qualified Crypto.PubKey.ECC.DH       as ECDH
import qualified Crypto.PubKey.ECC.ECDSA    as ECDSA
import qualified Crypto.PubKey.ECC.P256     as P256
import qualified Crypto.PubKey.ECC.Types    as ECC
import qualified Crypto.PubKey.ECC.Types    as ECCTypes
import           Data.Aeson                 ((.=))
import qualified Data.Aeson                 as A
import           Data.Bifunctor
import qualified Data.Binary                as Binary
import qualified Data.Bits                  as Bits
import qualified Data.ByteArray             as ByteArray
import           Data.ByteString            (ByteString)
import qualified Data.ByteString            as BS
import qualified Data.ByteString.Base64.URL as B64.URL
import qualified Data.ByteString.Lazy       as LB
import           Data.Data
import           Data.Text                  (Text)
import           Data.Word                  (Word16, Word64, Word8)
import           GHC.Int                    (Int64)

type VAPIDKeys = ECDSA.KeyPair

data PushNotificationPayload = PushNotificationPayload {
  payloadAudience :: Text
, payloadExpiration :: Int
, payloadSubject :: Text
} deriving (Show)

instance A.ToJSON PushNotificationPayload where
  toJSON p = A.object [
      "aud" .= payloadAudience p
    , "exp" .= payloadExpiration p
    , "sub" .= payloadSubject p
    ]

----------------------------
-- Manual implementation without using the JWT libraries.
-- Not using jose library. Check the below link for reason:
-- https://github.com/sarthakbagaria/web-push/pull/1#issuecomment-471254455
webPushJWT :: MonadIO m => VAPIDKeys -> PushNotificationPayload -> m LB.ByteString
webPushJWT vapidKeys payload = do
  let
    encodedJWTPayload = B64.URL.encodeBase64Unpadded' . LB.toStrict . A.encode $ payload
    messageForJWTSignature = encodedJWTHeader <> "." <> encodedJWTPayload
  -- JWT only accepts SHA256 hash with ECDSA for ES256 signed token
  -- ECDSA signing vulnerable to timing attacks
  ECDSA.Signature signR signS <- liftIO $ ECDSA.sign (ECDSA.toPrivateKey vapidKeys) SHA256 messageForJWTSignature
  -- 32 bytes of R followed by 32 bytes of S
  let encodedJWTSignature = B64.URL.encodeBase64Unpadded' $ LB.toStrict $ (Binary.encode $ int32Bytes signR) <> (Binary.encode $ int32Bytes signS)
  pure . LB.fromStrict $ messageForJWTSignature <> "." <> encodedJWTSignature
  where
    encodedJWTHeader = B64.URL.encodeBase64Unpadded' . LB.toStrict . A.encode $ A.object [
        "typ" .= ("JWT" :: Text)
      , "alg" .= ("ES256" :: Text)
      ]

-- | All inputs are in raw bytes with no encoding
-- except for the plaintext for which raw bytes are the Base 64 encoded bytes
data WebPushEncryptionInput = EncryptionInput {
  applicationServerPrivateKey :: ECDH.PrivateNumber
, userAgentPublicKeyBytes :: ByteString
, authenticationSecret :: ByteString
, salt :: ByteString
, plainText :: LB.ByteString
, paddingLength :: Int64
} deriving (Show)

-- | Intermediate encryption output used in tests
-- All in raw bytes
data WebPushEncryptionOutput = EncryptionOutput {
  sharedECDHSecretBytes :: ByteString
, inputKeyingMaterialBytes :: ByteString
, contentEncryptionKeyContext :: ByteString
, contentEncryptionKey :: ByteString
, nonceContext :: ByteString
, nonce :: ByteString
, paddedPlainText :: ByteString
, encryptedMessage :: ByteString
}

data PushEncryptError =
    PushEncryptCryptoError CryptoError
  | PushEncryptParseKeyError CryptoError
  | PushEncodeApplicationPublicKeyError String
  deriving (Eq, Show)

-- | payload encryption
-- https://tools.ietf.org/html/draft-ietf-webpush-encryption-04
webPushEncrypt :: WebPushEncryptionInput -> Either PushEncryptError WebPushEncryptionOutput
webPushEncrypt EncryptionInput {..} = do
  userAgentPublicKey <- first PushEncryptParseKeyError $ ecBytesToPublicKey userAgentPublicKeyBytes
  applicationServerPublicKeyBytes <- first PushEncodeApplicationPublicKeyError $ ecPublicKeyToBytes $ ECDH.calculatePublic curveP256 applicationServerPrivateKey
  let 
    sharedECDHSecret = ECDH.getShared curveP256 applicationServerPrivateKey userAgentPublicKey

    -- HMAC key derivation (HKDF, here expanded into HMAC steps as specified in web push encryption spec)
    pseudoRandomKeyCombine = HMAC.hmac authenticationSecret sharedECDHSecret :: HMAC.HMAC SHA256
    authInfo = "Content-Encoding: auth" <> "\x00" :: ByteString
    inputKeyingMaterial = HMAC.hmac pseudoRandomKeyCombine (authInfo <> "\x01") :: HMAC.HMAC SHA256

    context = "P-256" <> "\x00" <> "\x00" <> "\x41" <> userAgentPublicKeyBytes <> "\x00" <> "\x41" <> applicationServerPublicKeyBytes

    pseudoRandomKeyEncryption = HMAC.hmac salt inputKeyingMaterial :: HMAC.HMAC SHA256
    contentEncryptionKeyContext = "Content-Encoding: aesgcm" <> "\x00" <> context
    contentEncryptionKey = BS.pack $ take 16 $ ByteArray.unpack (HMAC.hmac pseudoRandomKeyEncryption (contentEncryptionKeyContext <> "\x01") :: HMAC.HMAC SHA256)

    nonceContext = "Content-Encoding: nonce" <> "\x00" <> context
    nonce = BS.pack $ take 12 $ ByteArray.unpack (HMAC.hmac pseudoRandomKeyEncryption (nonceContext <> "\x01") :: HMAC.HMAC SHA256)

    -- HMAC a doesn't have Show instance needed for test suite
    -- so we extract the bytes and store that in WebPushEncryptionOutput
    inputKeyingMaterialBytes = ByteArray.convert inputKeyingMaterial
    sharedECDHSecretBytes = ByteArray.convert sharedECDHSecret

    -- padding length encoded in 2 bytes, followed by
    -- padding length times '0' byte, followed by
    -- message
    paddedPlainText = LB.toStrict $
                          (Binary.encode (fromIntegral paddingLength :: Word16)) <>
                          (LB.replicate paddingLength (0 :: Word8)) <>
                          plainText

  -- aes_gcm is aead (authenticated encryption with associated data)
  -- use cek as the encryption key and nonce as the initialization vector
  aesCipher :: AES128 <- handleCryptoError $ Cipher.cipherInit contentEncryptionKey
  aeadGcmCipher <- handleCryptoError $ Cipher.aeadInit Cipher.AEAD_GCM aesCipher nonce
  -- tag length 16 bytes (maximum), anything less than 16 bytes may not be secure enough
  -- spec says final encrypted size is 16 bits longer than the padded text
  -- NOTE: the final encrypted message must be sent as raw binary data
  let
    (authTagBytes, cipherText) = Cipher.aeadSimpleEncrypt aeadGcmCipher BS.empty paddedPlainText 16
    authTag = ByteArray.convert $ Cipher.unAuthTag authTagBytes
    encryptedMessage = cipherText <> authTag
  pure $ EncryptionOutput {..}
  where
    handleCryptoError = first PushEncryptCryptoError . eitherCryptoError
    curveP256 = ECCTypes.getCurveByName ECCTypes.SEC_p256r1

-- Conversions among integers and bytes
-- The bytes are in network/big endian order.
 {-
    -- DON'T use DER encoding to extract integer bytes
    -- if a 32 byte number can be written in less bytes with leading zeros,
    -- DER encdoing will be shorter than 32 bytes
    -- and decoding to 4 word64 will fail because of short input
 -}
ecPublicKeyToBytes :: ECC.Point -> Either String ByteString
ecPublicKeyToBytes p = Crypto.ECC.encodePoint (Proxy :: Proxy Crypto.ECC.Curve_P256R1) <$> fromECCPoint p
  where
    fromECCPoint ECC.PointO = Left "Invalid public key infinity point"
    fromECCPoint (ECC.Point x y) = Right $ P256.pointFromIntegers (x,y)

ecBytesToPublicKey :: ByteString -> Either CryptoError ECC.Point
ecBytesToPublicKey = eitherCryptoError . fmap toECCPoint . Crypto.ECC.decodePoint (Proxy :: Proxy Crypto.ECC.Curve_P256R1)
  where toECCPoint = uncurry ECC.Point . P256.pointToIntegers 

-- Coordinates on Elliptic Curves are 32 bit integers
type Bytes32 = (Word64, Word64, Word64, Word64)

 -- points on elliptic curve for 256 bit algorithms are 32 bytes (256 bits) unsigned integers each
 -- fixed width big endian format
 -- integer to 4 Word64 (8 bytes each)
int32Bytes :: Integer -> Bytes32
int32Bytes number =  let shift1 = Bits.shiftR number 64
                         shift2 = Bits.shiftR shift1 64
                         shift3 = Bits.shiftR shift2 64
                     in ( fromIntegral shift3
                        , fromIntegral shift2
                        , fromIntegral shift1
                        , fromIntegral number
                       )

bytes32Int :: Bytes32 -> Integer
bytes32Int (d,c,b,a) = (Bits.shiftL (fromIntegral d) (64*3)) +
                       (Bits.shiftL (fromIntegral c) (64*2)) +
                       (Bits.shiftL (fromIntegral b) (64  )) +
                                    (fromIntegral a)


web-push-example/src/WebPushExample.hs
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE EmptyDataDecls        #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings     #-}
{-# LANGUAGE TypeOperators         #-}
module WebPushExample where


import           Control.Concurrent.STM.TVar
import           Control.Lens
import           Control.Monad
import           Control.Monad.IO.Class
import           Control.Monad.STM
import           Control.Monad.Trans.Reader
import qualified Data.Aeson                  as JS
import qualified Data.ByteString             as BS
import qualified Data.ByteString.Lazy        as BSL
import           Data.Set                    (Set)
import qualified Data.Set                    as Set
import           Data.Text
import qualified Data.Text                   as T
import qualified Data.Text.Encoding          as TE
import           Data.Time
import           Data.Word
import qualified Network.HTTP.Client         as HTTP
import qualified Network.HTTP.Conduit        as HTTP
import           Network.HTTP.Media          ((//), (/:))
import           Network.Wai.Handler.Warp
import           Servant
import           Servant.API
import           Servant.Server.StaticFiles
import           System.Directory
import qualified Text.Mustache               as Mustache
import           Web.FormUrlEncoded
import qualified Web.WebPush                 as WP
import           Web.WebPush                 (sendPushNotification)

import           Templates

-- | Form for sending a push notification, just the text to send
newtype PushNotificationForm = PushNotificationForm { pushNotificationText :: Text }

instance FromForm PushNotificationForm where
  fromForm form = PushNotificationForm <$> parseUnique "text" form

-- | Subscription to push notifications from the browser
data Subscription = Subscription {
  subEndpoint :: Text
, subAuth     :: Text
, subP256dh   :: Text
} deriving (Eq, Ord, Show)

instance JS.FromJSON Subscription where
  parseJSON = JS.withObject "Subscription" $ \obj -> Subscription
    <$> obj JS..: "endpoint"
    <*> obj JS..: "auth"
    <*> obj JS..: "p256dh"

instance JS.ToJSON Subscription where
  toJSON (Subscription endpoint auth p256dh) = JS.object [
      "endpoint" JS..= endpoint
    , "auth" JS..= auth
    , "p256dh" JS..= p256dh
    ]

instance FromForm Subscription where
  fromForm form =
    Subscription
      <$> parseUnique "endpoint" form
      <*> parseUnique "auth" form
      <*> parseUnique "p256dh" form

-- Server API that serves the the static files, subscribes user agents to push notifications, and sends push notifications
type API =
       "send" :> ReqBody '[FormUrlEncoded] PushNotificationForm :> Post '[JSON] () -- ^ Send a push notification to all subscribers
  :<|> "subscribe" :> ReqBody '[FormUrlEncoded] Subscription :> Post '[JSON] () -- ^ Subscribe to notifications
  :<|> Raw -- ^ Serves the static files (index.html, index.js, and service-worker.js) for the browser

-- | Configuration for the web server
data AppConfig = AppConfig
  { appConfigVAPIDKeys            :: WP.VAPIDKeys -- ^ Public and private keys for web-push
  , appConfigVAPIDKeyBytes        :: [Word8] -- ^ Public key for web-push as bytes, passed into the index.js.mustache template
  , appConfigManager              :: HTTP.Manager -- ^ HTTP manager for sending push notifications
  , appConfigSubscriptions        :: TVar (Set Subscription) -- ^ Active subscriptions to push notifications
  , appConfigWriteSubscriptions   :: IO () -- ^ Write subscriptions to a file
  }

-- | Handler for the API, reads the config from the environment
type PushHandler = ReaderT AppConfig Handler

server :: [Word8] -> ServerT API PushHandler
server vapidPublicKey =
       postSendPushNotification
  :<|> postAddSubscriber
  :<|> serveDirectoryEmbedded staticFiles
  where
    -- The static files to serve to the browser
    -- `index.js` has the VAPID public key embedded in it
    staticFiles = [
          ("service-worker.js", serviceWorkerJS )
        , ("index.html", indexHtml)
        , ("index.js", TE.encodeUtf8 $ Mustache.substituteValue indexJsTemplate $ Mustache.object [ "serverKey" Mustache.~> vapidPublicKey ])
      ]

-- | Add a subscriber to the list of subscribers
postAddSubscriber :: Subscription -> PushHandler ()
postAddSubscriber sub = do
  writeSubscriptions <- asks appConfigWriteSubscriptions
  subscriptionRef <- asks appConfigSubscriptions
  liftIO $ do
    putStrLn "Adding subscriber"
    atomically $ modifyTVar' subscriptionRef (Set.insert sub)
    writeSubscriptions

-- | Filepath subscriptions are persisted to
subscriptionsFp :: FilePath
subscriptionsFp = "subscriptions.json"

-- | Write all subscriptions to a file
writeSubscriptions :: TVar (Set Subscription) -> IO ()
writeSubscriptions ref =  do
  subs <- atomically $ readTVar ref
  JS.encodeFile subscriptionsFp subs

-- | Initialize subscriptions from a file if they exist
initPersistentSubscriptions :: IO (TVar (Set Subscription))
initPersistentSubscriptions = do
  e <- doesFileExist subscriptionsFp
  if e
    then do
      subs <- JS.decodeFileStrict subscriptionsFp
      case subs of
        Nothing   -> fail "Unable to read subscriptions"
        Just subs -> newTVarIO subs
    else newTVarIO mempty

-- | Send a push notification to all subscribers
postSendPushNotification :: PushNotificationForm -> PushHandler ()
postSendPushNotification (PushNotificationForm text) = do
  cfg <- ask
  liftIO $ appSendPushNotification cfg text

-- | Send a push notification to all subscribers 
appSendPushNotification :: AppConfig -- ^ Configuration for the application
                        -> Text -- ^ Text to send in the push notification
                        -> IO ()
appSendPushNotification cfg text = do
  -- Get the current time to use as the tag for the notification
  time <- getCurrentTime
  -- Get all the subscriptions
  subscribtions <- atomically $ readTVar subs
  let
    -- This is the message sent to browsers
    message = JS.object [
          "title" JS..= ("Web Push Test" :: Text)
        , "body" JS..= text
        , "icon" JS..= ("" :: Text)
        , "tag" JS..= T.pack (show time)
        , "url" JS..= ("http://localhost:3000" :: Text)
        ]
    pushDetails (Subscription endpoint auth p256dh) =
      (WP.mkPushNotification endpoint p256dh auth) -- The subscription details
        & WP.pushExpireInSeconds .~ 60 * 60 * 12 -- 12 hours
        & WP.pushMessage .~ message -- The message we created above
        & WP.pushSenderEmail .~ "test@example.com" -- The email address of the sender

  liftIO $ putStrLn $ "Sending notification to " <> show (Prelude.length subscribtions) <> " subscribers containing message: " <> show text
  -- Loop over the subscriptions sending each one a notification
  forM_ subscribtions $ \sub -> do
    liftIO $ putStrLn $ "Sending notification to: " <> show sub
    notificationResult <- WP.sendPushNotification keys manager $ pushDetails sub
    liftIO $ putStrLn $ "Notification result: " <> show notificationResult
  where
    keys = appConfigVAPIDKeys cfg
    manager = appConfigManager cfg
    subs = appConfigSubscriptions cfg

runExampleApp :: Int -- ^ Port to run the application on
              -> AppConfig -- ^ Configuration for the application
              -> IO ()
runExampleApp port cfg = run port (serveWithContextT (Proxy :: Proxy API) EmptyContext (flip runReaderT cfg) (server (appConfigVAPIDKeyBytes cfg)))

initInMemoryConfig :: IO AppConfig
initInMemoryConfig = do
  vapidKeys <- either fail (pure . WP.readVAPIDKeys) =<< WP.generateVAPIDKeys
  subscriptions <- newTVarIO mempty
  keyBytes <- either fail pure $ WP.vapidPublicKeyBytes vapidKeys
  manager <- HTTP.newManager HTTP.tlsManagerSettings
  pure $ AppConfig vapidKeys keyBytes manager subscriptions (pure ())

initPersistentConfig :: IO AppConfig
initPersistentConfig = do
  vapidKeys <- initPersistentKeys
  subscriptions <- initPersistentSubscriptions
  keyBytes <- either fail pure $ WP.vapidPublicKeyBytes vapidKeys
  manager <- HTTP.newManager HTTP.tlsManagerSettings
  pure $ AppConfig vapidKeys keyBytes manager subscriptions (writeSubscriptions subscriptions)
  where
    initPersistentKeys :: IO WP.VAPIDKeys
    initPersistentKeys = do
      keysExist <- doesFileExist fp
      if keysExist
        then (maybe (fail "Unable to read keys") (pure . WP.readVAPIDKeys . from)) =<< JS.decodeFileStrict fp
        else do
          putStrLn "Generating keys"
          keys <- either fail pure =<< WP.generateVAPIDKeys
          JS.encodeFile fp $ to keys
          pure $ WP.readVAPIDKeys keys
      where
        fp = "vapid-keys.json"
        to (WP.VAPIDKeysMinDetails n x y) = (n , x , y)
        from (n , x , y) = WP.VAPIDKeysMinDetails n x y
    -- Initialize subscriptions from a file if they exist
    initPersistentSubscriptions :: IO (TVar (Set Subscription))
    initPersistentSubscriptions = do
      e <- doesFileExist subscriptionsFp
      if e
        then do
          subs <- JS.decodeFileStrict subscriptionsFp
          case subs of
            Nothing   -> fail "Unable to read subscriptions"
            Just subs -> newTVarIO subs
        else newTVarIO mempty
    -- Write all subscriptions to a file
    writeSubscriptions :: TVar (Set Subscription) -> IO ()
    writeSubscriptions ref =  do
      subs <- atomically $ readTVar ref
      JS.encodeFile subscriptionsFp subs